### ISSUE TITLE
ODBC: "Unexpected driver error occurred while connecting to the database"

### ISSUE THREAD

**Issue by marcreig** [dbeaver/dbeaver#12457]
Good afternoon,

I've recently bought a new server (big tower with an Intel Xeon inside).
I've installed a Windows 10 PRO today and updated to 20H2 version.
Installed Java JRE x86 and x64 last versión (java version "1.8.0_291") and finally, installed DBeaver (version 21.0.4, Community Ed).
This software configuration, is the very same I've in a Notebook, and on my office PC. Even on my Surface Pro 6.
We work on a company which our ERP software DB is a 4D SQL engine (v15.6).
So we've the specific 4D ODBC driver (v17) which we install on windows and then, let's us create an ODBC link on Windows ODBC Data Source administrator. For particular case, I use x64 version of driver and windows administrator.
We create a User DSN and it connect perfectly well, as it does on ALL mentioned  previous machines:

https://i.imgur.com/bh137A9.png

Then on DBeaver, I've only to create a simple ODBC new connection and write the DSN name, as I always do.
But in this case, I obtain an error: "Unexpected driver error occurred while connecting to the database"

https://i.imgur.com/ZPlh2PG.png

while in all other matchines it works with no problem and I can connect:

https://i.imgur.com/vk4c2aE.png

I've tried EVERYTHING. Unisntalled all javas, tried JDKs versions. Tried v7 versions. Donwloaded different sources of drivers for DBeaver. Even copied project folders from machines working ones, to the server one.
Even read about BRIDGE unsuported on java 8 (which I don't understant because it works perfectly well on other machines) and tried diferent solutions found on stackoverflow & Co.
Nothing. It doesn't connects. And on the log, it only says:

2021-05-13 19:05:10.440 - org.jkiss.dbeaver.model.exec.DBCConnectException: Unexpected driver error occurred while connecting to the database
org.jkiss.dbeaver.model.exec.DBCConnectException: Unexpected driver error occurred while connecting to the database
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:227)
	at org.jkiss.dbeaver.ext.generic.model.GenericDataSource.openConnection(GenericDataSource.java:135)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCExecutionContext.connect(JDBCExecutionContext.java:101)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCRemoteInstance.initializeMainContext(JDBCRemoteInstance.java:100)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCRemoteInstance.<init>(JDBCRemoteInstance.java:59)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.initializeRemoteInstance(JDBCDataSource.java:109)
	at org.jkiss.dbeaver.ext.generic.model.GenericDataSource.<init>(GenericDataSource.java:107)
	at org.jkiss.dbeaver.ext.generic.model.meta.GenericMetaModel.createDataSourceImpl(GenericMetaModel.java:90)
	at org.jkiss.dbeaver.ext.generic.GenericDataSourceProvider.openDataSource(GenericDataSourceProvider.java:106)
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:897)
	at org.jkiss.dbeaver.runtime.jobs.ConnectJob.run(ConnectJob.java:70)
	at org.jkiss.dbeaver.runtime.jobs.ConnectionTestJob.run(ConnectionTestJob.java:103)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:105)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.lang.NullPointerException
	at sun.jdbc.odbc.JdbcOdbcDriver.initialize(JdbcOdbcDriver.java:453)
	at sun.jdbc.odbc.JdbcOdbcDriver.connect(JdbcOdbcDriver.java:153)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.lambda$0(JDBCDataSource.java:176)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:195)
	... 13 more
2021-05-13 19:05:10.506 - 
java.lang.NullPointerException
	at sun.jdbc.odbc.JdbcOdbcDriver.initialize(JdbcOdbcDriver.java:453)
	at sun.jdbc.odbc.JdbcOdbcDriver.connect(JdbcOdbcDriver.java:153)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.lambda$0(JDBCDataSource.java:176)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:195)
	at org.jkiss.dbeaver.ext.generic.model.GenericDataSource.openConnection(GenericDataSource.java:135)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCExecutionContext.connect(JDBCExecutionContext.java:101)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCRemoteInstance.initializeMainContext(JDBCRemoteInstance.java:100)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCRemoteInstance.<init>(JDBCRemoteInstance.java:59)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.initializeRemoteInstance(JDBCDataSource.java:109)
	at org.jkiss.dbeaver.ext.generic.model.GenericDataSource.<init>(GenericDataSource.java:107)
	at org.jkiss.dbeaver.ext.generic.model.meta.GenericMetaModel.createDataSourceImpl(GenericMetaModel.java:90)
	at org.jkiss.dbeaver.ext.generic.GenericDataSourceProvider.openDataSource(GenericDataSourceProvider.java:106)
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:897)
	at org.jkiss.dbeaver.runtime.jobs.ConnectJob.run(ConnectJob.java:70)
	at org.jkiss.dbeaver.runtime.jobs.ConnectionTestJob.run(ConnectionTestJob.java:103)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:105)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
> Start Núcleo del equipo de soporte [org.eclipse.team.core 3.8.1100.v20200806-0621]
2021-05-13 19:10:03.725 - DBeaver is stopping
2021-05-13 19:10:03.726 - Stop RMI server
2021-05-13 19:10:03.727 - DBeaver shutdown completed

Any idea? I don't know what more to do.
Frustrating that it perfectly works on other machines! I've been working more than a year on this way!

Thanks and regards.

**Comment by p91paul** [comment]
https://community.oracle.com/tech/developers/discussion/1344007/nullpointerexception-in-jdbcodbcdriver-initialize-jdbcodbcdriver-java-435
this guy says it solved it by setting some permissions on some odbc dll. however they are not very specific....

**Comment by marcreig** [comment]
Anyone? I'm still on this problem, and it doesn't seem to have a logical solution...
About permisions, it doesn't make any sense, as it is a common installation as I've already done on every other computer where I have the environtment correctly working...

**Comment by p91paul** [comment]
https://stackoverflow.com/questions/14229072/removal-of-jdbc-odbc-bridge-in-java-8
See this. In particular, did you put the jdbcodbc.dll file in the correct location?
Also keep in mind that by default Dbeaver will not use your installed java version, instead it will run on the one bundled with the installation program, you'll have to instruct it to use another jre. Also, the claim you are running on jdk 8 looks odd to me, I believed recent versions of dbeaver required jdk 11.
So make sure which jre dbeaver is actually running on, and that the jdbcodbc bridge is properly installed in that version.

**Comment by marcreig** [comment]
Hi,

drivers are on it's default location ("userX/AppData/Roaming/DBeaverData/drivers/..."). Where they get installed on inmediate download from manager, when you try to create the very first connection after a fresh DBeaver installation.

Not a permissions problem. Tried to create User DSN and System DSN on ODBC Windows manager.
In both profiles: x64 and x86 drivers (which I got).
Nothing...
I delete everything. Uninstall every part and repeat the process. Nothing. Rinse and repeat.
It is like DBeaver is not even trying to connect. Like you push the TEST button and it just throws the error...
Works fine on Postgre or MariaDB or Oracle connections. The problem is just on ODBC ones.

And I need to repeat it: I do exactly the very same I do on all the other computers and environtments! All on Windows 10 x64, some in Pro, some in Home. But anyway working correctly.
Don't know what more to watch for...

Looked at installation folder on Program Files, and JRE is bundled as you said. But compared multiple installation versions and they're all the same: JAVA_VERSION="11.0.9.1"

So no clue on what to do.
What I can say is, that all my Python scripts with PYODBC work perfectly on that new "server". So is not a problem of ODBC 4Dsql drivers or their installations. Not even their configurations. Python works flawlessly. It connects, it queryies, it receives.

It looks like an strict matter of DBeaver or java. 
Is there a way to test ODBC (JDBC) connection without DBeaver just to verify its java or its only DBeaver?

Thanks.
